File: solutions/csharp/kindergarten-garden/1/KindergartenGarden.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Plant { Violets, Radishes, Clover, Grass }

public class KindergartenGarden
{
    private enum Student { Alice, Bob, Charlie, David, Eve, Fred, Ginny, Harriet, Ileana, Joseph, Kincaid, Larry };
    private readonly Dictionary<Student, IEnumerable<Plant>> _plantsByStudent = new Dictionary<Student, IEnumerable<Plant>>();

    public KindergartenGarden(string diagram)
    {
        _plantsByStudent = diagram
          .Split('\n')
          .SelectMany(row => Enumerable.Range(0, row.Length / 2).Select((integer, position) => (Student: (Student)position, Plants: row.Substring(integer * 2, 2).Select(GetPlant))))
          .GroupBy(t => t.Student, t => t.Plants)
          .Select(g => (g.Key, g.SelectMany(i => i)))
          .ToDictionary(t => t.Item1, t => t.Item2);
    }

    private Plant GetPlant(char plant) => plant switch
    {
        'R' => Plant.Radishes,
        'C' => Plant.Clover,
        'G' => Plant.Grass,
        'V' => Plant.Violets,
        _ => throw new ArgumentException("Invalid plant type specified")
    };

    public IEnumerable<Plant> Plants(string student) => _plantsByStudent[(Student)Enum.Parse(typeof(Student), student)];
}

File: solutions/fsharp/kindergarten-garden/1/KindergartenGarden.fs
module KindergartenGarden

type Plant =
    | Radishes
    | Clover
    | Grass
    | Violets

let students =
    [ "Alice"
      "Bob"
      "Charlie"
      "David"
      "Eve"
      "Fred"
      "Ginny"
      "Harriet"
      "Ileana"
      "Joseph"
      "Kincaid"
      "Larry" ]

let getPlantByInitial =
    function
    | 'R' -> Radishes
    | 'C' -> Clover
    | 'G' -> Grass
    | 'V' -> Violets
    | _ -> failwith "Invalid plant type specified"

let plants (diagram: string) (student: string) =

    let fstIdx =
        List.findIndex (fun i -> i = student) students * 2
    let sndIdx = fstIdx + 1

    let row1 =
        diagram.Split([| '\n' |]).[0] |> List.ofSeq
    let row2 =
        diagram.Split([| '\n' |]).[1] |> List.ofSeq

    [ row1.[fstIdx]
      row1.[sndIdx]
      row2.[fstIdx]
      row2.[sndIdx] ]
    |> List.map getPlantByInitial